repose-aggregator/components/filters/saml-policy-translation-filter/src/main/scala/org/openrepose/filters/samlpolicy/SamlPolicyTranslationFilter.scala

```diff
@@ -21,21 +21,184 @@
-  extends AbstractConfiguredFilter[SamlPolicyConfig](configurationService){
+class SamlPolicyTranslationFilter @Inject()(configurationService: ConfigurationService,
+                                            atomFeedService: AtomFeedService,
+                                            akkaServiceClientFactory: AkkaServiceClientFactory)
+  extends AbstractConfiguredFilter[SamlPolicyConfig](configurationService)
+  with LazyLogging
+  with AtomFeedListener {
 
   override val DEFAULT_CONFIG: String = "saml-policy.cfg.xml"
   override val SCHEMA_LOCATION: String = "/META-INF/config/schema/saml-policy.xsd"
 
-  override def doWork(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = chain.doFilter(request, response)
+  private var cache: LoadingCache[String, XsltExecutable] = _
+
+  override def doWork(servletRequest: ServletRequest, servletResponse: ServletResponse, chain: FilterChain): Unit = {
+    try {
+      val request = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+      var response = servletResponse.asInstanceOf[HttpServletResponse]
+      val samlResponse = decodeResponse(request)
+      val rawDocument = readToDom(samlResponse)
+
+      val version = determineVersion(rawDocument)
+      val finalDocument = version match {
+        case 1 =>
+          request.addHeader("Identity-API-Version", "1.0")
+          rawDocument
+        case 2 =>
+          response = new HttpServletResponseWrapper(response, ResponseMode.PASSTHROUGH, ResponseMode.MUTABLE)
+          request.addHeader("Identity-API-Version", "2.0")
+          val issuer = validateResponseAndGetIssuer(rawDocument)
+          val translatedDocument = translateResponse(rawDocument, cache.get(issuer))
+          signResponse(translatedDocument)
+      }
+      val inputStream = convertDocumentToStream(finalDocument)
+
+      request.replaceHeader(CONTENT_TYPE.toString, MediaType.APPLICATION_XML)
+
+      chain.doFilter(new HttpServletRequestWrapper(request, inputStream), response)
+
+      if (version == 2) {
+        doResponseStuff()
+        response.asInstanceOf[HttpServletResponseWrapper].commitToResponse()
+      }
+    } catch {
+      case ex: SamlPolicyException =>
+        servletResponse.asInstanceOf[HttpServletResponse].sendError(ex.statusCode, ex.message)
+        logger.debug("SAML policy translation failed", ex)
+      case ex: Exception =>
+        servletResponse.asInstanceOf[HttpServletResponse].sendError(SC_INTERNAL_SERVER_ERROR, "Unknown error in SAML Policy Filter")
+        logger.warn("Unexpected problem in SAML Policy Filter", ex)
+    }
+  }
+
+  /**
+    * Gets the SAMl response from the post encoded body, and decodes it into a plain String.
+    *
+    * @param request the servlet request that has an encoded body
+    * @return the decoded saml response
+    * @throws SamlPolicyException if decoding fails
+    */
+  def decodeResponse(request: HttpServletRequest): String = ???
```

Comment:
I'm good either way, i just assumed you had a string coming out of the methods you were calling